quantum/process_keycode/process_key_lock.c

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2017 Fredric Silberberg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "inttypes.h"
+#include "stdint.h"
+#include "process_key_lock.h"
+
+#define SHIFT(shift) (((uint64_t)1) << (shift))
```

Comment:
I've implemented a lot of the functionality with macros, as I wanted to have very few function calls for this. If this isn't the preferred style, I can move this into separate functions and suggest to the compiler that it can inline them if it would like to.